src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3326,29 +3326,48 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                     x.base.base.loc);
             }
         }
-        LCOMPILERS_ASSERT(
-            ASRUtils::check_equal_type(ASRUtils::expr_type(lhs), ASRUtils::expr_type(rhs)));
+        ASR::ttype_t *left_operand_type = ASRUtils::expr_type(lhs);
+        ASR::ttype_t *right_operand_type = ASRUtils::expr_type(rhs);
+
         ASR::expr_t *value = nullptr;
-        ASR::ttype_t *dest_type = ASRUtils::expr_type(lhs);
+        ASR::ttype_t *dest_type = left_operand_type;
 
+        if (ASR::is_a<ASR::Character_t>(*left_operand_type)) {
+            throw SemanticError("Logical operation not supported on object of type 'str'", lhs->base.loc);
+        }
+        if (ASR::is_a<ASR::Character_t>(*right_operand_type)) {
+            throw SemanticError("Logical operation not supported on object of type 'str'", rhs->base.loc);
+        }
         if (ASRUtils::expr_value(lhs) != nullptr && ASRUtils::expr_value(rhs) != nullptr) {
```

Comment:
Inside this `if` we compute the the compile-time value. Currently it computes only for logical types. We need to extend support for other types.

Reply:
Can be done for sure! But what about the back-end? Will we be removing it?